File: 15.三数之和.java
/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        
        for(int i = 0; i < nums.length -2; i++){
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            int l = i + 1;
            int r = nums.length - 1;
            while(l < r){
                int result = nums[i] + nums[l] + nums[r];
                if(result == 0){
                    res.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    while(l < r && nums[l] == nums[l + 1]) l ++;
                    while(l < r && nums[r] == nums[r - 1]) r --;
                    l++;
                    r--;
                }else if (result > 0){
                    r = r - 1;
                }else{
                    l = l + 1;
                }
            }
        }
        return res;
    }
}
// @lc code=end

File: 15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	var res = make([][]int, 0)
	n := len(nums)
	sort.Ints(nums)
	
	for i := 0; i < n - 2; i++ {
		if i > 0 && nums[i] == nums[i - 1] {
			continue;
		}
		left := i + 1
		right := n - 1
		
		for left < right {
			sum := nums[i] + nums[left] + nums[right]
			if sum == 0 {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				for left < right && nums[left] == nums[left + 1] {
					left ++
				}
				for left < right && nums[right] == nums[right - 1] {
					right --
				}
				left ++
				right --
			} else if sum > 0{
				right --
			} else {
				left ++
			}
		}
	}
	return res
}
// @lc code=end

